pkg/metrics/metrics_test.go

```
@@ -0,0 +1,346 @@
+	testVersion   = "1.0.0"
+	testInstallID = "yv66vt6tvu8="
+	testServerURL = "https://example.com"
+)
+
+func defaultClient() *client {
+	return &client{
+		AppID:      testAppID,
+		Version:    testVersion,
+		InstallID:  testInstallID,
+		HTTPClient: &http.Client{Timeout: 1 * time.Second},
+		OptOut:     false,
+		Config: &metricsConfig{
+			ServerURL: testServerURL,
+			NoMetrics: false,
+		},
+	}
+}
+
+func TestNew(t *testing.T) {
+	t.Parallel()
+	t.Run("happy_path", func(t *testing.T) {
+		t.Parallel()
+
+		cases := []struct {
+			name      string
+			client    *http.Client
+			installID string
+			want      *client
+		}{
+			{
+				name: "happy_path_no_install_id",
+				want: defaultClient(),
+			},
+			{
+				name:      "happy_path_with_install_id",
+				installID: testInstallID,
+				want:      defaultClient(),
+			},
+			{
+				name:      "happy_path_with_custom_http_client",
+				installID: testInstallID,
+				client:    &http.Client{Timeout: 2},
+				want: func() *client {
+					c := defaultClient()
+					c.HTTPClient = &http.Client{Timeout: 2}
+					return c
+				}(),
+			},
+		}
+
+		for _, tc := range cases {
+			t.Run(tc.name, func(t *testing.T) {
+				t.Parallel()
+
+				ctx := context.Background()
+
+				installPath := t.TempDir() + "/" + installIDFileName
+				if tc.installID != "" {
+					if err := storeInstallID(testAppID, installPath, &InstallIDData{tc.installID}); err != nil {
+						t.Fatalf("test setup failed: %s", err.Error())
+					}
+				}
+				envVars := map[string]string{
+					"METRICS_URL": testServerURL,
+				}
+				lookuper := envconfig.MapLookuper(envVars)
+				opts := []Option{
+					WithLookuper(lookuper),
+					WithInstallIDFileOverride(installPath),
+				}
+				if tc.client != nil {
+					opts = append(opts, WithHTTPClient(tc.client))
+				}
+
+				i, err := New(ctx, testAppID, testVersion, opts...)
+				if err != nil {
+					t.Errorf("unexpected error: %s", err.Error())
+				}
+				got := i.(*client) //nolint:forcetypeassert
```

It'd probably be better to do the type check

```suggestion
				got, ok := i.(*client) //nolint:forcetypeassert
				if !ok {
				    t.Fatal(...)
				}
```

---

I see, I guess I figured I would just let it panic, but I'll change those.

---

Left one in testcase definition as it would no longer be an easy one-liner (would need a closure or something)